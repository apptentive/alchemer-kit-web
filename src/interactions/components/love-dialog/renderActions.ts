type EventHandler = (event: Event) => void;

/**
 * Renders a given Love Dialog's set of actions.
 *
 * @param {string} yesButtonText - The text to render for the 'Love' button.
 * @param {Function} yesClickHandler - The action to take when the 'Love' button is clicked.
 * @param {string} noButtonText - The text to render for the 'Do Not Love' button.
 * @param {Function} noClickHandler - The action to take when the 'Do Not Love' button is clicked.
 * @returns {HTMLDivElement} - The actions DOM Node.
 */
const renderActions = (
  yesButtonText: string,
  yesClickHandler: EventHandler,
  noButtonText: string,
  noClickHandler: EventHandler
) => {
  const actions = document.createElement('div');
  actions.className = 'apptentive-love-dialog-actions';

  // Yes
  const yesAction = document.createElement('button');
  yesAction.textContent = yesButtonText || 'Yes';
  yesAction.classList.add('apptentive-love-dialog-action');
  yesAction.classList.add('apptentive-love-dialog-yes');
  yesAction.addEventListener('click', yesClickHandler);

  // No
  const noAction = document.createElement('button');
  noAction.textContent = noButtonText || 'No';
  noAction.classList.add('apptentive-love-dialog-action');
  noAction.classList.add('apptentive-love-dialog-no');
  noAction.addEventListener('click', noClickHandler);

  actions.append(noAction);
  actions.append(yesAction);

  return actions;
};

export default renderActions;
