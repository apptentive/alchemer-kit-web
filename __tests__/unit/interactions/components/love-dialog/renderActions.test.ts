import renderActions from '../../../../../src/interactions/components/love-dialog/renderActions';

const _yesButtonText = 'Love it!';
const _noButtonText = 'Hate it!';
const _yesClickHandler = jest.fn();
const _noClickHandler = jest.fn();

describe('renderActions', () => {
  afterEach(() => {
    _yesClickHandler.mockClear();
    _noClickHandler.mockClear();
  });

  test('properly renders actions', () => {
    const output = renderActions(_yesButtonText, _yesClickHandler, _noButtonText, _noClickHandler);

    expect(output).toMatchSnapshot();
  });

  test('properly renders without any options', () => {
    const output = renderActions(null as any, null as any, null as any, null as any);

    expect(output).toMatchSnapshot();
  });

  test('properly attaches click handlers', () => {
    document.body.append(renderActions(_yesButtonText, _yesClickHandler, _noButtonText, _noClickHandler));

    const yesButton = document.querySelector<HTMLButtonElement>('.apptentive-love-dialog-yes')!;
    const noButton = document.querySelector<HTMLButtonElement>('.apptentive-love-dialog-no')!;

    yesButton.click();
    noButton.click();

    expect(_yesClickHandler).toHaveBeenCalledTimes(1);
    expect(_noClickHandler).toHaveBeenCalledTimes(1);

    document.body.innerHTML = '';
  });
});
