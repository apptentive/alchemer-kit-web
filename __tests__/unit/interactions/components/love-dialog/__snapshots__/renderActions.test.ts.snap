// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderActions properly renders actions 1`] = `
<div
  class="apptentive-love-dialog-actions"
>
  <button
    class="apptentive-love-dialog-action apptentive-love-dialog-no"
  >
    Hate it!
  </button>
  <button
    class="apptentive-love-dialog-action apptentive-love-dialog-yes"
  >
    Love it!
  </button>
</div>
`;

exports[`renderActions properly renders without any options 1`] = `
<div
  class="apptentive-love-dialog-actions"
>
  <button
    class="apptentive-love-dialog-action apptentive-love-dialog-no"
  >
    No
  </button>
  <button
    class="apptentive-love-dialog-action apptentive-love-dialog-yes"
  >
    Yes
  </button>
</div>
`;
